### ISSUE TITLE
VSCode autoimport keeps adding import { null } from 'mathjs';

### ISSUE THREAD

**Issue by thunderkid** [josdejong/mathjs#2019]
I hope it makes sense to post this issue here - it really does seem specific to `mathjs`.

I'm using VSCode to edit a project that includes `mathjs` plus around 50 other libraries. After I've edited something I very often find that VSCode has inserted the line

`import { null } from 'mathjs';`

at the top of the file, which I then have to go and manually delete. Admittedly this only takes a few seconds, but since I have to repeat the process dozens of times per day, it becomes tedious.
This is coming from VSCode's auto-import feature, which I don't want to turn off since it's otherwise very useful. Of all the hundreds of packages that I've used over time, `mathjs` is the only one that I've ever found to do this. 
Is `mathjs` exporting things in some non-standard way that would cause this problem? Is there some workaround?

**Comment by josdejong** [comment]
Thanks for reporting @thunderkid, that is frustrating. 

I guess it is caused somehow because mathjs contains the value `null` on it's instance, like `math.null`, `math.true`, and `math.false`. This is only useful though when using those constants in the expression parser (that's why they are attached on the mathjs instance in first place) 

I guess we can safely remove them from the exported instance, we only need to make sure they are available in the expression parser. 

We have to figure out which export exactly does cause this autoimport in VSCode. Is this coming from the mathjs library itself or from it's type definitions?

**Comment by thunderkid** [comment]
Thanks for the answer. I don't know how to check whether the problem comes from the library or the type definitions. 

In my configuration `package.json` contains the lines `"@types/mathjs": "^6.0.5"` and also `"mathjs": "^7.2.0"`. Inside my `node_modules` folder I have a `mathjs` folder but no `@mathjs` folder. What should I remove to test this?

**Comment by josdejong** [comment]
I think it's coming from the TypeScript definitions. I created a project with mathjs and types, and choose the "autoimport" on `null` in VSCode. When I now Ctrl+click the `null`, I'm brought to the following line in the type definitions:

```js
interface MathJsStatic extends FactoryDependencies {
        // ...
        null: number;
        // ...
}
```

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cf707b6cfc440d5dd3a1d6824a63d413854001ad/types/mathjs/index.d.ts#L54

I think we should do two things: 

1. remove `null` from the TypeScript definition
2. do not expose `null`, `true`, and `false` on the mathjs instance, but only have them available inside the expression parser. We have to think that through though.

Anyone able to help out removing `null` from the TypeScript definitions?

**Comment by digimbyte** [comment]
just an update, I have this instance too but we're not using TypeScript
I use null values for state management and it keeps importing itself when I look away.
easy fix but it's getting old after 2 weeks

**Comment by josdejong** [comment]
Thanks for your input @digimbyte . Interesting to hear that this is also happening without TypeScript, though it may be that your IDE is using the TypeScript definitions under the hood anyway. What IDE are you using? VS Code?

Again: any help would be very welcome here, it shouldn't be too hard to create a PR removing `null` from the type definitions.

**Comment by digimbyte** [comment]
standard VS Code with "mathjs": "^9.2.0"
it must be some sort of auto-complete but have no idea where or why it
would be doing it.

**Comment by josdejong** [comment]
I wouldn't be surprised when VS Code would install type defininitions in the background when it finds them. A first easy step would be to see if removing the `null` from those type definitions fixes the issue already.

**Comment by wmr-trevor** [comment]
Just for a little more context here.

This is how I see VSCode showing that it'll auto import the null value.
![image](https://user-images.githubusercontent.com/78549186/115434495-21b1a000-a1df-11eb-9d68-fb225556f1c4.png)

Allowing it to import I see that the declaration is in `@types/mathjs/index.d.ts on line 54` in my case. `version 6.0.5`.

Removing that line in the declaration file prevents VsCode from auto importing that null value.

**Comment by josdejong** [comment]
Thanks @wmr-trevor , so fixing this is really a oneliner at DefinitelyTyped 👍 

Anyone able to create PR at DefinitelyTyped to fix this?

**Comment by keriwarr** [comment]
@josdejong ^

**Comment by josdejong** [comment]
😎  thanks Keri!

**Comment by josdejong** [comment]
Fix is now also published in `mathjs@9.4.0` which now contains type definitions itself, no need to fetch them from Definitely Typed anymore
